### DISCUSSION TITLE
start_tls_error on invalid password, invalid username, lockouts, expiry all working normal

### DISCUSSION THREAD

**Discussion by Narfmeister** [DirectoryTree/LdapRecord#702] [Q&A] [answered]
Hi, 

I'm currently attempting to implement ldapRecord and am having luck with authenticating and reading user details. 

What I'm encountering is when an invalid password is supplied I receive an LdapRecord \LdapRecordException of `ldap_start_tls(): Unable to start TLS: Local error.` What seems odd is that after 3 attempts, the account is locked out in ActiveDirectory and returns the correct lock out message. Something odd seems to be happening specifically when an invalid password is supplied.

Supplying an invalid username for example seems to appear as expected: 
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/d84758de-9925-497b-8481-a3a262d0b990)

I'm just following along with the Laravel Breeze guide in the documentation and I'm not sure where I've gone wrong as the authentication appears to be succeeding and other exceptions such as lockouts & invalid usernames appear to be processing correctly. It's only invalid passwords where this is happening. Has anyone experienced this or have any advice?

I'm currently using Laravel 10.45.0 with ldaprecord-laravel 3.4.0 and PHP 8.3.3

Cheers

**Comment by stevebauman** [comment]
Hi @Narfmeister,

Can you please post your connection configuration with any sensitive details omitted?

**Reply by Narfmeister** [reply]
Hi @stevebauman  

Sure thing:
```
LDAP_LOGGING=true
LDAP_CONNECTION=default
LDAP_HOST="xxx"
LDAP_USERNAME="xxx@xxx"
LDAP_PASSWORD="xxx"
LDAP_PORT=389
LDAP_BASE_DN="DC=xxx,DC=xxx"
LDAP_TIMEOUT=5
LDAP_SSL=false
LDAP_TLS=true
LDAP_SASL=false
```

A snippet from the trace is as follows if it might help with anything:
```
[previous exception] [object] (ErrorException(code: 2): ldap_start_tls(): Unable to start TLS: Local error at D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\Ldap.php:154)
[stacktrace]
#0 [internal function]: LdapRecord\\Ldap->LdapRecord\\{closure}(2, 'ldap_start_tls(...', 'D:\\\\bla\\\\...', 154)
#1 D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\Ldap.php(154): ldap_start_tls(Object(LDAP\\Connection))
#2 D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\HandlesConnection.php(161): LdapRecord\\Ldap->LdapRecord\\{closure}()
#3 D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\Ldap.php(153): LdapRecord\\Ldap->executeFailableOperation(Object(Closure))
#4 D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\Auth\\Guard.php(84): LdapRecord\\Ldap->startTLS()
#5 D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\Auth\\Guard.php(125): LdapRecord\\Auth\\Guard->bind('notadodgyusername...', 'totallyasecurepassword')
```

**Reply by stevebauman** [reply]
Thanks for posting that @Narfmeister. Everything looks fine there.

Have you tried the debugging steps here? If not, can you give it a shot and let me know the result?

https://ldaprecord.com/docs/laravel/v3/debugging/#tls--ssl

**Reply by Narfmeister** [reply]
No problems @stevebauman 
I gave them a go, `ldap:browse` connects successfully:
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/fb215bcd-d0b8-432e-b7d8-b1d9eed03cf5)


As does `ldap:test`:
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/5d9bb4ba-3ff9-40d9-9ad1-10ff0f18fd65)


I can login successfully and I have the listener added for my login controller already:
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/c110b91e-31dd-41a8-8f3a-b3c4ead4e2ed)

Invalid username: 
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/a8f3b846-839e-4875-842b-d30f47a6ceb1)

Invalid password:
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/87598ece-70d6-4070-8184-45b8d2327380)

When I do this 3 times and intentionally lock the account it then parses this exception correctly now:
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/945ca153-5e78-4a53-9c24-2e0b485b34c4)

If I unlock the acct and supply valid credentials, it succeeds as we expect: 
![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/25217203/62acafa2-3839-4d54-ac9e-34b61db2de01)

It seems to specifically be related to an invalid password being supplied and how it handles that exception maybe? It seems to be attempting to authenticate as I can see account locking after 3 attempts in AD.

**Reply by stevebauman** [reply]
Strange, can you catch the LdapRecord exception that occurs and dump the detailed error that is available as a method here?

https://github.com/DirectoryTree/LdapRecord/blob/5ff91b636748631796835fa1b6a8cd8c9c497611/src/LdapRecordException.php#L35

Post what is dumped here and we can debug.

**Reply by Narfmeister** [reply]
Below these logs are a stacktrace which I can post if necessary. I'm not sure how to access the output of `getDetailedError()` I'm sorry... Could I trouble you for some help on how I can get this info for you?

I've extracted this from the laravel logs which shows the invalid credential being received and the exception happening after:
```
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Binding - Username: service-account@my.server  
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Bound - Username: service-account@my.server  
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Search - Base DN: dc=myDomain - Filter: (&(objectclass=\74\6f\70)(objectclass=\70\65\72\73\6f\6e)(objectclass=\6f\72\67\61\6e\69\7a\61\74\69\6f\6e\61\6c\70\65\72\73\6f\6e)(objectclass=\75\73\65\72)(userprincipalname=\61\2d\70\68\69\70\70\73\6b\61\40\74\65\73\74\71\68\2e\63\6c\6f\75\64\2e\68\65\61\6c\74\68\2e\71\6c\64\2e\67\6f\76\2e\61\75)(!(objectclass=\63\6f\6d\70\75\74\65\72))) - Selected: (objectguid,*) - Time Elapsed: 53.03  
[2024-02-23 02:38:57] local.DEBUG: User [Narfmeister] has been successfully discovered for authentication.  
[2024-02-23 02:38:57] local.DEBUG: Object with name [Narfmeister] is being synchronized.  
[2024-02-23 02:38:57] local.DEBUG: Object with name [Narfmeister] has been successfully synchronized.  
[2024-02-23 02:38:57] local.DEBUG: User [Narfmeister] is authenticating.  
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Attempting - Username: CN=Narfmeister,OU=Accounts,dc=myDomain  
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Binding - Username: CN=Narfmeister,OU=Accounts,dc=myDomain  
[2024-02-23 02:38:57] local.WARNING: LDAP (ldap://my.server:389) - Operation: Failed - Username: CN=Narfmeister,OU=Accounts,dc=myDomain - Reason: Invalid credentials  
[2024-02-23 02:38:57] local.INFO: LDAP (ldap://my.server:389) - Operation: Binding - Username: service-account@my.server  
[2024-02-23 02:38:57] local.ERROR: ldap_start_tls(): Unable to start TLS: Local error {"exception":"[object] (LdapRecord\\LdapRecordException(code: 2): ldap_start_tls(): Unable to start TLS: Local error at D:\\bla\\vendor\\directorytree\\ldaprecord\\src\\LdapRecordException.php:19)
```

**Reply by stevebauman** [reply]
No worries -- we'll need the exception details, not just the exception message, as it provides additional error codes to indicate what the issue is with the server.

If I recall correctly, in Laravel Breeze you should have a `app/Http/Requests/Auth/LoginRequest.php` where you can catch the exception, or you can simply CTRL+F to locate anywhere in your application where `Auth::attempt` is being called:

```php
// app/Http/Requests/Auth/LoginRequest.php

// ...

public function authenticate(): void
{
    $this->ensureIsNotRateLimited();

    try {
        if (! Auth::attempt($this->only('email', 'password'), $this->boolean('remember'))) {
            RateLimiter::hit($this->throttleKey());
        
            throw ValidationException::withMessages([
                'email' => trans('auth.failed'),
            ]);
        }
    
        RateLimiter::clear($this->throttleKey());
    } catch (\LdapRecord\LdapRecordException $e) {
        dd($e->getDetailedError());
    }
}
```

**Reply by Narfmeister** [reply]
Thank you for that, I wasn't getting the `catch` part right it seems!

Here's the output from that for you: 
```
LdapRecord\DetailedError {#1415 ▼ // app\Http\Requests\Auth\LoginRequest.php:57
  #errorCode: 49
  #errorMessage: "Invalid credentials"
  #diagnosticMessage: "80090308: LdapErr: DSID-0C090447, comment: AcceptSecurityContext error, data 52e, v3839"
}
```

**Reply by Narfmeister** [reply]
Hey @stevebauman just following up on this. Was the information I posted above what you were looking for? 

Cheers

**Reply by jollyrogers11** [reply]
Hey @stevebauman, I am also seeing this issue. I am updating some projects of mine to Laravel 11 and it is requiring me to update directorytree/ldaprecord-laravel. They were at version 2.7 and I have updated them to 3.3.1 and had not seen any issues when testing locally on my machine so I pushed my changes into production. I was monitoring my log files on a few sites that I had updated and saw the following:

[2024-03-26 08:59:39] production.ERROR: ldap_start_tls(): Unable to start TLS: Local error {"exception":"[object] (LdapRecord\\LdapRecordException(code: 2): ldap_start_tls(): Unable to start TLS: Local error

I wasn't sure what was causing the issue because I seemed to be able to login to the site just fine. Started doing some searches to see if anyone else was having the issue and found this forum. I tried logging in with an invalid password and it throws this exception:

![image](https://github.com/DirectoryTree/LdapRecord-Laravel/assets/97048634/71110837-29ce-4906-9db8-1990ed5997b6)

It only seems to happen when I try logging in with a valid username but invalid password. 

My authentication controller is pretty simple. I am not using an Eloquent model, I just check to see if the authentication attempt with the LDAP credentials is successful or not:

public function authenticate(Request $request)
{
        $validated = $request->validate([
            'username' => 'required',
            'password' => 'required',
        ]);

        Event::listen(Rejected::class, function () {
            abort(redirect()->back()->withErrors(['message' => 'You are unauthorized to access this page.']));
        });

        if (!Auth::attempt(['samaccountname' => $validated['username'], 'password' => $validated['password']])) {
            return redirect()->back()->withErrors(['message' => 'Username or password is incorrect.']);
        }

        return redirect(route('home'));
}

I do have an LDAP rule that makes sure you are in certain groups and should be able to login. I have always just done the Auth::attempt and if it returns false, I would redirect back with an error message. This is crashing for me now because an LdapRecordException is being thrown when I attempt to login with an invalid password. Should that exception be thrown for invalid password or should it just return false? I have about 30 sites that I maintain. Do I need to update them all to just catch the LdapRecordException and then redirect with an error message? Any thoughts on the best way to fix this issue are much appreciated.

**Reply by stevebauman** [reply]
Hey guys apologies for the delay on this, I'll be able to look into this over the next couple days. I finally have a Windows VM with Active Directory in Azure so I can ensure this is fixed. Will keep you posted, but of course PR's are open if you find a resolution sooner 🙏

**Reply by jollyrogers11** [reply]
I started doing incremental updates this morning to see when the issue shows up for me. Everything works as expected up to version 2.7.3, but I start getting the issue when I update to 3.0.0.

**Reply by stevebauman** [reply, marked as answer]
Hi guys, this issue should be resolved in the latest update [v3.6.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v3.6.0). I appreciate your patience.

Please run `composer update` and you should be all set.

Let me know if you encounter any issues! 🙏

**Reply by Narfmeister** [reply]
Works perfectly! Thank you!
![image](https://github.com/DirectoryTree/LdapRecord/assets/25217203/7f9b1b3c-e9c7-445b-807b-cad563ad8dae)

**Reply by stevebauman** [reply]
Awesome! Thanks for testing and reporting back 🙏

**Reply by jollyrogers11** [reply]
Updated this morning and my sites are working again like they used to. Thanks @stevebauman !!

**Reply by stevebauman** [reply]
Great to hear @jollyrogers11! Thanks for reporting 🙏
